sormas-app/app/src/main/java/de/symeda/sormas/app/component/dialog/SynchronizationDialog.java

```diff
@@ -0,0 +1,585 @@
+
+		currentBinding = null;
+		currentSyncStep = synchronizationStep;
+
+		switch (synchronizationStep) {
+		case PUSH_NEW:
+			showPushNewProgressItems();
+			break;
+		case PULL_INFRASTRUCTURE:
+			showPullInfrastructureProgressItems(false);
+			break;
+		case PULL_INFRASTRUCTURE_INITIAL:
+			showPullInfrastructureProgressItems(true);
+			break;
+		case CLEAR_UP:
+			showClearUpProgressItems(true);
+			break;
+		case SYNCHRONIZE:
+			showSynchronizeProgressItems();
+			break;
+		case CLEAR_UP_INFRASTRUCTURE:
+			showClearUpInfrastructureProgressItems(true);
+			break;
+		case DELETE_OBSOLETE:
+			showDeleteObsoleteProgressItems();
+			break;
+		case PULL_MODIFIED:
+			showPullModifiedProgressItems();
+			break;
+		case REPULL:
+			showRepullProgressItems();
+			break;
+		default:
+			throw new IllegalArgumentException(synchronizationStep.name());
+		}
+
+		getActivity().runOnUiThread(() -> {
+			contentBinding.stepsLayout.setCurrentStep(++currentSyncStepIndex);
+			contentBinding.stepsLayout.stepText
+				.setText(String.format(getActivity().getResources().getString(synchronizationStep.getCaptionResource()), currentSyncStepIndex));
+		});
+	}
+
+	private void updateStepNumber(int stepNumber) {
+		getActivity().runOnUiThread(() -> contentBinding.stepsLayout.setStepNumber(stepNumber));
+	}
+
+	private void loadNext() {
+
+		waitForProgressItemBindings();
+
+		currentBinding.setActive(false);
+		currentBinding.setDone(true);
+
+		if (progressItemBindings.size() > ++progressItemBindingsIndex) {
+			currentBinding = progressItemBindings.get(progressItemBindingsIndex);
+			currentBinding.setActive(true);
+		}
+	}
+
+	private void updatePulls(int pulled) {
+		waitForProgressItemBindings();
+		currentBinding.setPulls(currentBinding.getPulls() + pulled);
+	}
+
+	private void updatePushes(int pushed) {
+		waitForProgressItemBindings();
+		currentBinding.setPushes(currentBinding.getPushes() + pushed);
+	}
+
+	private void updatePushTotal(int pushTotal) {
+		waitForProgressItemBindings();
+		currentBinding.setPushTotal(pushTotal);
+	}
+
+	private void updateDeletions(int deleted) {
+		waitForProgressItemBindings();
+		currentBinding.setDeletions(currentBinding.getDeletions() + deleted);
+	}
+
+	private void showNextCleanupItems() {
+
+		currentBinding = null;
+
+		if (currentSyncStep == SynchronizationStep.CLEAR_UP_INFRASTRUCTURE) {
+			showClearUpInfrastructureProgressItems(false);
+		} else if (currentSyncStep == SynchronizationStep.CLEAR_UP) {
+			showClearUpProgressItems(false);
+		} else {
+			throw new UnsupportedOperationException("Current sync step is no clear up step");
+		}
+	}
+
+	private void showProgressItems(boolean showPulls, boolean showPushes, List<String> captions) {
+
+		showProgressItems(showPulls, showPushes, false, captions);
+	}
+
+	private void showProgressItems(boolean showPulls, boolean showPushes, boolean showDeletions, List<String> captions) {
+
+		getActivity().runOnUiThread(() -> {
+			assert (currentBinding == null);
```

Comment:
This lead to an error on my system, which made me aware of the following situation:

When updateSynchronizationStep is called twice, before the UI thread is executed again, it will mean that the logic in showProgressItems is also executed twice, once the UI thread is run. For the second execution the currentBinding will not be null.

Unfortunately simply setting currentBinding to null here will also not solve the problem. The first UI thread execution for showProgressItems will notify the waiting calls that access the currentBinding. If the second exeuction of showProgressItems is running in parallel it will take away the currentBinding and either lead to wrong progress data or worst case a nullpointerexception. @MateStrysewske Let's discuss how we can mitigate this.